internal/app/orchestrator/orchestrator.go

```diff
@@ -95,16 +114,186 @@ func New(ctx context.Context,
+			}
+		}
+	}
+
+	return responseChannel, o.onCancelWatch(aggregatedKey, &req)
 }
 
 // Fetch implements the polling method of the config cache using a non-empty request.
-func (c *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
+func (o *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
 	return nil, fmt.Errorf("Not implemented")
 }
 
-func (c *orchestrator) onCacheEvicted(key string, resource cache.Resource) {
-	// TODO implement.
+// watchResponse is intended to be called in a goroutine, to receive incoming
+// responses and fan out to downstream clients.
+func (o *orchestrator) watchUpstream(
+	ctx context.Context,
+	aggregatedKey string,
+	responseChannel <-chan *discovery.DiscoveryResponse,
+	done <-chan bool,
+	shutdown func(),
+) {
+	for {
+		select {
+		case x, more := <-responseChannel:
+			if !more {
+				// A problem occurred fetching the response upstream, log retry.
+				// TODO implement retry/back-off logic on error scenario.
+				// https://github.com/envoyproxy/xds-relay/issues/68
+				o.logger.With("key", aggregatedKey).Error(ctx, "upstream error")
+				return
+			}
+			// Cache the response.
+			_, err := o.cache.SetResponse(aggregatedKey, *x)
+			if err != nil {
+				// TODO if set fails, we may need to retry upstream as well.
+				// Currently the fallback is to rely on a future response, but
+				// that probably isn't ideal.
+				// https://github.com/envoyproxy/xds-relay/issues/70
+				//
+				// If we fail to cache the new response, log and return the old one.
+				o.logger.With("err", err).With("key", aggregatedKey).
+					Error(ctx, "Failed to cache the response")
+			}
+
+			// Get downstream watches and fan out.
+			// We retrieve from cache rather than directly fanning out the
+			// newly received response because the cache does additional
+			// resource serialization.
+			cached, err := o.cache.Fetch(aggregatedKey)
+			if err != nil {
+				o.logger.With("err", err).With("key", aggregatedKey).Error(ctx, "cache fetch failed")
+				// Can't do anything because we don't know who the watches
+				// are. Drop the response.
+			} else {
+				if cached == nil || cached.Resp == nil {
+					// If cache is empty, there is nothing to fan out.
+					// Error. Sanity check. Shouldn't ever reach this since we
+					// just set the response, but it's a rare scenario that can
+					// happen if the cache TTL is set very short.
+					o.logger.With("key", aggregatedKey).Error(ctx, "attempted to fan out with no cached response")
+				} else {
+					// Goldenpath.
+					o.fanout(cached.Resp, cached.Requests)
+				}
+			}
+		case <-done:
+			// Exit when signaled that the stream has closed.
+			shutdown()
+			return
+		}
+	}
+}
+
+// fanout pushes the response to the response channels of all open downstream
+// watches in parallel.
+func (o *orchestrator) fanout(resp *cache.Response, watchers map[*gcp.Request]bool) {
+	for watch := range watchers {
```

Comment:
probably need to make a copy and then iterate?

Reply:
Are you worried about cache ttl removing the watches in the middle of the fanout? I'm afraid we can't make a copy as the downstreamResponseMap requires a pointer to the watch, in order to retrieve the channel.

Reply:
there can be multiple situations in which the watch could be changed.  a watch could be removed when sidecar disconnects and also on ttl expiry. Im not sure how iterators in java work. in java changing a iterator while changing it will cause exception.

Reply:
This should be fine in Go: https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop

Reply:
when the watch is deleted, the channel will be gone too. We will be pushing into a closed channel, isnt it?

Reply:
Synced offline. Our decision is to not close the downstream channels (on cache evict and on cancel). Garbage collection will clean up and close the channels. This way we don't have to worry about pushing to a closed channel in a different routine.

Reply:
> Is there a reason why the map is a pointer of the request rather than value?

Yes, because value isn't sufficient to get the corresponding downstream watch. Cache was originally by value, and I was not able to retrieve the same watch so we moved all these to pointers.